cdc/log.cc

```diff
@@ -539,12 +539,109 @@ class transformer final {
+                return type.compare(p.first, k) >= 0;
+            });
+            if (i != res.end() && i->first == k) {
+                i->second = v;
+            } else {
+                res.emplace(i, k, v);
+            }
+        });
+        set_for_each(deleted, [&res](bytes_view k) {
+            auto i = std::find_if(res.begin(), res.end(), [&k](std::pair<bytes_view, bytes_view>& p) {
+                return p.first == k;
+            });
+            if (i != res.end()) {
+                res.erase(i);
+            }
+        });
+        return map_type_impl::serialize_partially_deserialized_form(res, cql_serialization_format::internal());
+    }
+    static bytes merge_sets(const abstract_type& type, const bytes_opt& prev, const bytes_opt& next, const bytes_opt& deleted) {
+        std::vector<bytes_view> res;
+        set_for_each(prev, [&res](bytes_view k) {
+            res.emplace_back(k);
+        });
+        set_for_each(next, [&res, &type](bytes_view k) {
+            auto i = std::find_if(res.begin(), res.end(), [&k, &type](bytes_view rk) {
+                return type.compare(rk, k) >= 0;
+            });
+            if (i == res.end() || *i != k) {
+                res.emplace(i, k);
+            }
+        });
+        set_for_each(deleted, [&res](bytes_view k) {
+            auto i = std::find_if(res.begin(), res.end(), [&k](bytes_view rk) {
+                return rk == k;
+            });
+            if (i != res.end()) {
+                res.erase(i);
+            }
+        });
+        return set_type_impl::serialize_partially_deserialized_form(res, cql_serialization_format::internal());
+    }
+    static bytes merge_udt(const user_type_impl& type, const bytes_opt& prev, const bytes_opt& next, const bytes_opt& deleted) {
+        std::vector<bytes_view_opt> res(type.size());
+        udt_for_each(prev, [&res, i = res.begin()](bytes_view_opt k) mutable {
+            *i++ = k;
+        });
+        udt_for_each(next, [&res, i = res.begin()](bytes_view_opt k) mutable {
+            if (k) {
+                *i = k;
+            }
+            ++i;
+        });
+        set_for_each(deleted, [&res](bytes_view k) {
+            auto index = deserialize_field_index(k);
+            res[index] = std::nullopt;
+        });
+        return type.build_value(res);
+    }
+
     // TODO: is pre-image data based on query enough. We only have actual column data. Do we need
     // more details like tombstones/ttl? Probably not but keep in mind.
     mutation transform(const mutation& m, const cql3::untyped_result_set* rs, api::timestamp_type ts, bytes tuuid, int& batch_no) const {
         auto stream_id = _ctx._cdc_metadata.get_stream(ts, m.token());
         mutation res(_log_schema, stream_id.to_partition_key(*_log_schema));
-
+        // should we generate post-image data? note: we can only do this if both pre- and postimage 
```

Comment:
It's probably better for postimage=true to implicate preimage=true and check it when the user set's that. It can be done in separate patch though.

Reply:
I don't agree. Pre-image will add more data to the log table. If the user only wants post-image, he will have to pay the read-before-write price, but he should not be forced to suffer more log space usage than requested.

Reply:
Makes sense.